### ISSUE TITLE
Sprout doesn't time out SESSION_TERMINATED ASs fast enough

### ISSUE THREAD

**Issue by mirw** [Metaswitch/sprout#1018]
> I have Project Clearwater and a 2-site AS, fronted by a pool of 3 proxies.
> I have a SIPp script running call load – 400 callers and 400 callees.
> All callers have originating services on site 1 with DefaultHandling SESSION_TERMINATED.
> All callees have terminating services on site 2 with DefaultHandling SESSION_TERMINATED.
> If I take down the first proxy in the pool, Sprout tries to use it and declares it dead after 4 seconds.
> **It then takes ~30s for Sprout to try the second proxy in the pool.**
> This proxy forwards the call on to the correct AS, which sends back a 100 Trying.
> Sprout rejects the 100 Trying, sending a Cancel, presumably because it has already torn down the call in those 30s.

After 4s, our "liveness timer" pops and we generate an internal 408 response but this doesn't trigger a retry:

```
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:89: Sproutlet timer popped, id = 140447578697408
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1492: Timer has popped
24-04-2015 11:31:17.475 UTC Verbose sproutletproxy.cpp:1259: scscf-0x7fbc801b8b10 cancelling fork 0, reason = 0
24-04-2015 11:31:17.475 UTC Debug pjutils.cpp:697: Cloned tdta0x7fbc804cbf20 to tdta0x7fbc5c559000
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1067: Remove top Route header Route: <sip:odi_phufA66FgU@10.248.0.50:5054;lr;orig>
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1500: Adding message 0x7fbc5c559610 => txdata 0x7fbc5c5590a8 mapping
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1500: Adding message 0x7fbc5d19c850 => txdata 0x7fbc5d19c2e8 mapping
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1507: Removing message 0x7fbc5c559610 => txdata 0x7fbc5c5590a8 mapping
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1295: Free message tdta0x7fbc5c559000
24-04-2015 11:31:17.475 UTC Debug pjsip: tdta0x7fbc5c55 Destroying txdata Request msg INVITE/cseq=1 (tdta0x7fbc5c559000)
24-04-2015 11:31:17.475 UTC Verbose sproutletproxy.cpp:1218: scscf-0x7fbc801b8b10 sending Response msg 408/INVITE/cseq=1 (tdta0x7fbc5d19c240)
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1515: Processing actions from sproutlet - 1 responses, 0 requests
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1590: Aggregating response with status code 408
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1640: 3xx/4xx/5xx/6xx response
24-04-2015 11:31:17.475 UTC Debug sproutletproxy.cpp:1644: Best 3xx/4xx/5xx/6xx response so far
24-04-2015 11:31:17.475 UTC Verbose sproutletproxy.cpp:1560: scscf-0x7fbc801b8b10 fork 0 pending CANCEL, state = Calling
```

After 32s, the transaction times out - this is the point that we retry.

```
24-04-2015 11:31:45.476 UTC Debug pjsip: tsx0x7fbc804d0 Timeout timer event
24-04-2015 11:31:45.476 UTC Debug pjsip: tsx0x7fbc804d0 State changed from Calling to Terminated, event=TIMER
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:214: tsx0x7fbc804d0ef8 - tu_on_tsx_state UAC, TSX_STATE TIMER state=Terminated
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:1803: tsx0x7fbc804d0ef8 - uac_tsx = 0x7fbc801422d0, uas_tsx = 0x7fbc8046d260
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:1811: TIMER event on current UAC transaction
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:2116: Stopping timer C
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:1833: Failed to connected to server or timed-out, so add to blacklist
24-04-2015 11:31:45.476 UTC Debug baseresolver.cpp:461: Add ##.##.##.##:5060 transport 6 to blacklist for 30 seconds
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:1982: Attempt to retry request to alternate server
24-04-2015 11:31:45.476 UTC Debug pjsip: tsx0x7fbc5c364 Transaction created for Request msg INVITE/cseq=1 (tdta0x7fbc5d198a10)
24-04-2015 11:31:45.476 UTC Debug basicproxy.cpp:2009: Created transaction for retry, so send request
24-04-2015 11:31:45.476 UTC Debug pjsip: tsx0x7fbc5c364 Sending Request msg INVITE/cseq=1 (tdta0x7fbc5d198a10) in state Null
24-04-2015 11:31:45.476 UTC Debug pjsip:       endpoint Request msg INVITE/cseq=1 (tdta0x7fbc5d198a10): skipping target resolution because address is already set
```

So the internally-generated 408 response doesn't seem to be having any effect.

There's also a question here about whether we should retry these requests - we're timing out at the application level after 4s, which I think means we won't retry at the transport level.  Should we be retrying at the transport level faster than this?

**Comment by mirw** [comment]
I had a look at whether PJSIP had built-in control of TCP-level connection timeouts (which are relevant to the last point) - it looks like it doesn't.

The stack here is `lis_create_transport` calls `pj_activesock_start_connect` which calls `pj_ioqueue_connect`.  There's a question about which layer we should time TCP connections out at.  `lis_create_transport` is probably easiest, but it's not very generic.  That's probably OK, because PJSIP is only used for SIP (not the other protocols that could be supported by the same I/O layer.

We could call `pjsip_endpt_schedule_timer` to set a timer and then call `pj_activesock_close` in the timer callback,

**Comment by mirw** [comment]
Bumping to high priority as we need to get this fixed.

**Comment by AlexHockey** [comment]
We have now implemented the following:

* A timeout on TCP connection establishment (which fires if the SYN-ACK is not received promptly). 
* A timeout on TCP message transfer (which fires if a PSH ACK is not received promptly). 

Both of these are configurable but default to 2s. Both timeouts result in a transport failure being reported up to sprout which fails any transactions in progress on that connection (and causes the remote host to be blacklisted). 

This means that a failed ASs will be timed out promptly and we can retry to a working AS within the session-terminated timeout. The timeout is now configurable as well, so can be increased / disabled completely if needed. 

Note that session-terminated processing is currently broken in other ways but I've raised this as a separate issue (see https://github.com/Metaswitch/sprout/issues/1080).
